### PR TITLE
add CRF tutorial

### PR DESCRIPTION
**Pull request by howl-anderson**
# Description

Brief Description of the PR:

add a tutorial for CRF Layer

**To Reviewers**:

Here is the candidate feature list for this tutorial:
* [Done] Replace faked training data with real training data:
   * Pros: make the tutorial more practicable
   * Cons: Add many codes for data processing, make it longer and less concentration.
* [Done] Add additional custom training step base approach (co-exists with current custom training loop based approach):
   * Pros: User have more choose of programming style
   * Cons: Add some (not much) codes, make it longer and less concentration.

What do you think? 

Fixes # (issue)

## Type of change

- [ ] Bug fix
- [x] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [ ] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [ ] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

If you're adding a bugfix or new feature please describe the tests that you ran to verify your changes:
*

### CODE REVIEW

**Comment by howl-anderson** [comment]
Update: I am working on update this tutorial with real training data.

**Comment by howl-anderson** [comment]
Update: all the planed features are implemented. it still needs some polish but it's ready for technical review.

**Comment by howl-anderson** [comment]
When #2555 is merged, I will update this tutorial accrodingly.

**Comment by bhack** [comment]
@howl-anderson Can you join in the official forum? 

We have some user looking for CRF support:
https://discuss.tensorflow.org/t/how-to-implement-and-use-a-linear-chain-crf-in-tensorflow/4742/

**Comment by howl-anderson** [comment]
@bhack Sure, I will handle this. Thank you!

**Comment by howl-anderson** [comment]
Update: since #2555 is merged, I will update this tutorial soon.

**Comment by howl-anderson** [comment]
@seanpmorgan Please take a look

File: docs/tutorials/layers_crf.ipynb [hunk 1]
```diff
@@ -0,0 +1,1203 @@
...
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "xHxb-dlhMIzW"
+      },
+      "source": [
+        "## Overview\n",
+        "\n",
+        "This notebook will demonstrate how to use the CRF (Conditional Random Field) layer in TensorFlow Addons.\n",
+        "\n",
+        "You will learn how to use the CRF layer in two ways by building NER models."
```
Comment (MarkDaoust) [inline docs/tutorials/layers_crf.ipynb:68]: Please add a bit more description on CRF and NER.

What do these do? How do I know if they're the right tool for me?

File: docs/tutorials/layers_crf.ipynb [hunk 2]
```diff
@@ -0,0 +1,1203 @@
...
+      "metadata": {
+        "id": "U82B_tH2d294"
+      },
+      "outputs": [],
+      "source": [
+        "optimizer = tf.keras.optimizers.Adam(0.02)\n",
+        "\n",
+        "train_loss = tf.keras.metrics.Mean(name=\"train_loss\")\n",
+        "\n",
+        "@tf.function(experimental_relax_shapes=True)\n",
```
Comment (MarkDaoust) [inline docs/tutorials/layers_crf.ipynb:784]: SO that wrapper class mainly just inserts this as the model.train_step, right?

**Review by MarkDaoust — CHANGES_REQUESTED** [review]
This looks like a pretty solid start.

I don't know much about this topic so my comments are mostly just style suggestions.

Anything you can do to flesh-out the descriptions would be helpful for anyone who ends up here without significant background.

File: docs/tutorials/layers_crf.ipynb [hunk 3]
```diff
@@ -0,0 +1,1203 @@
...
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "rEk-ibQkDNtF"
+      },
+      "outputs": [],
+      "source": [
+        "!pip install -q tensorflow-addons  # version >= 0.15.0 is required\n",
+        "!pip install -q tensorflow\n",
+        "!pip install -q datasets"
```
[the remark this answers is not in the document]
Comment (howl-anderson) [reply docs/tutorials/layers_crf.ipynb:90]: Great suggestion! I will fix it in the next commit.

File: docs/tutorials/layers_crf.ipynb [hunk 4]
```diff
@@ -0,0 +1,1203 @@
...
+        "## Traning data"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "3db2a84b25a7"
+      },
+      "source": [
+        "Loading the CoNLL 2003 dataset by using the datasets library."
```
[the remark this answers is not in the document]
Comment (howl-anderson) [reply docs/tutorials/layers_crf.ipynb:124]: Great suggestion! I will fix it in the next commit.

File: docs/tutorials/layers_crf.ipynb [hunk 5]
```diff
@@ -0,0 +1,1203 @@
...
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "5WctVm30TFsT"
+      },
+      "outputs": [
```
[the remark this answers is not in the document]
Comment (howl-anderson) [reply docs/tutorials/layers_crf.ipynb:133]: I will fix it in the next commit.

File: docs/tutorials/layers_crf.ipynb [hunk 6]
```diff
@@ -0,0 +1,1203 @@
...
+          "text": [
+            "['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'British', 'lamb', '.']\n",
+            "['B-ORG', 'O', 'B-MISC', 'O', 'O', 'O', 'B-MISC', 'O', 'O']\n"
+          ]
+        }
+      ],
+      "source": [
+        "sample_tags = [raw_tags[i] for i in sample_tag_ids]\n",
+        "\n",
+        "print(sample_tokens)\n",
```
[the remark this answers is not in the document]
Comment (howl-anderson) [reply docs/tutorials/layers_crf.ipynb:452]: Great suggestion! I will fix it in the next commit.

File: docs/tutorials/layers_crf.ipynb [hunk 7]
```diff
@@ -0,0 +1,1203 @@
...
+        ")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "oPjyvdqaMEhU"
+      },
+      "source": [
+        "## Method one: Using the CRF layer in a custom training loop"
```
[the remark this answers is not in the document]
Comment (howl-anderson) [reply docs/tutorials/layers_crf.ipynb:641]: OK. I will fix it in the next commit.

File: docs/tutorials/layers_crf.ipynb [hunk 8]
```diff
@@ -0,0 +1,1203 @@
...
+        "      yield item['tokens'], item['ner_tags']\n",
+        "  \n",
+        "  return data_generator\n",
+        "\n",
+        "data_signature= (\n",
+        "        tf.TensorSpec(shape=(None,), dtype=tf.string),\n",
+        "        tf.TensorSpec(shape=(None, ), dtype=tf.int32)\n",
+        ")\n",
+        "\n",
+        "train_data = tf.data.Dataset.from_generator(\n",
```
[the remark this answers is not in the document]
Comment (howl-anderson) [reply docs/tutorials/layers_crf.ipynb:589]: I am not sure doing shuffle is a good idea. Since shuffle comes with randomness, it may make the model result unpredictable. On the other hand, it's a tutorial on CRF, and I am trying to make things simple. But if you consist that we need do a shuffle, that is OK. I can fix it in the following commit.

File: docs/tutorials/layers_crf.ipynb [hunk 9]
```diff
@@ -0,0 +1,1203 @@
...
+        "    tokens = tf.strings.lower(tokens)\n",
+        "    return lookup_layer(tokens)\n",
+        "\n",
+        "BATCH_SIZE = 2048\n",
+        "\n",
+        "# With `padded_batch()`, each batch may have different length\n",
+        "# shape: (batch_size, None)\n",
+        "train_dataset = (\n",
+        "    train_data.map(dataset_preprocess)\n",
+        "    .padded_batch(batch_size=BATCH_SIZE).cache()\n",
```
[the remark this answers is not in the document]
Comment (howl-anderson) [reply docs/tutorials/layers_crf.ipynb:631]: Same as the previous comment, I am trying to avoid making the model result unpredictable. I prefer to keep everything simple.

File: docs/tutorials/layers_crf.ipynb [hunk 10]
```diff
@@ -0,0 +1,1203 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "3b38225d9464"
+      },
+      "outputs": [],
+      "source": [
+        "@tf.function\n",
```
[the remark this answers is not in the document]
Comment (howl-anderson) [reply docs/tutorials/layers_crf.ipynb:751]: Fixed in the next commit.

**Comment by howl-anderson** [comment]
@MarkDaoust Please take a look

**Review by MarkDaoust — CHANGES_REQUESTED** [review]
LGTM.

Last couple of suggestions.

1. In the [notebook view](https://github.com/tensorflow/addons/blob/e6a31d21ad837d2ad6a02dd1da83e81549ad2db1/docs/tutorials/layers_crf.ipynb) There are a bunch of places where you have a full URL in the text, please move these to markdown links like:

```
See [this tutorial](https://url.url/url) for more information
```

2. The spots where you link to API pages please just put the symbol path in backticks. Replace: https://www.tensorflow.org/addons/api_docs/python/tfa/text/CRFModelWrapper with `tfa.text.CRFModelWrapper`. On tensorflow.org that will auto-link to the right page.

3. There are a couple of spots where you name symbols in the test, without backticks. Please add the backticks. Example: tfa.text.crf_log_likelihood should be `tfa.text.crf_log_likelihood`.

4. TFDS now has a `huggingface` adaptor. And it has the `conll2003` dataset available, thatcould really simplify a lot of the data-loading code.

https://www.tensorflow.org/datasets/community_catalog/huggingface?hl=en&%3Bskip_cache=true&skip_cache=true#conll2003

**Comment by seanpmorgan** [comment]
Thank you for your contribution. We sincerely apologize for any delay in reviewing, but TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
